File: algorithms/python/_259_3SumSmaller/_259_3SumSmaller.py
class Solution1:
    """
    @param nums:  an array of n integers
    @param target: a target
    @return: the number of index triplets satisfy the condition nums[i] + nums[j] + nums[k] < target
    """
    def threeSumSmaller(self, nums, target):
        # Write your code here
        list.sort(nums)
        sum = 0
        for i in range(len(nums)-2):
            lo, hi = i+1, len(nums)-1
            while lo < hi:
                while lo < hi and nums[lo] + nums[hi] >= target - nums[i]:
                    hi -= 1
                sum += hi - lo
                lo += 1
        return sum

class Solution:
    """
    @param nums:  an array of n integers
    @param target: a target
    @return: the number of index triplets satisfy the condition nums[i] + nums[j] + nums[k] < target
    """
    def threeSumSmaller(self, nums, target):
        # Write your code here
        nums.sort()
        n = len(nums)
        res = n*(n-1)*(n-2)//6
        for i in range(len(nums)-2):
            j = i+1;k=len(nums)-1
            while j < k:
                s = nums[i] + nums[j] + nums[k]
                #[i,j,k],[i,j+1,k],...,[i,k-1,k] all satisfy the condition
                if s >= target:
                    res -= k-j
                    k -= 1
                else:
                    j += 1
        return res

File: algorithms/cpp/_259_3SumSmaller/_259_3SumSmaller.cpp

# include <iostream>

using namespace std;

int main() {
    cout << "new file!" << endl;
    return 0;
}

class Solution {
public:
    /**
     * @param nums:  an array of n integers
     * @param target: a target
     * @return: the number of index triplets satisfy the condition nums[i] + nums[j] + nums[k] < target
     */
    int threeSumSmaller(vector<int> &nums, int target) {
        // Write your code here
        sort(nums.begin(),nums.end());
        int ans = 0;
        int len = nums.size();
        for( int i = 0;i < len - 2; ++i )
        {
            int sum = target - nums[i];
            int start =  i + 1;
            int end = len - 1;
            while( start < end )
            {
                if( nums[start] + nums[end] >= sum )
                {
                    --end;
                }else{
                    ans += (end - start);
                    ++start;
                }
            }
        }
        return ans;
    }
};

class Solution1 {
public:
    int threeSumSmaller(vector<int>& nums, int target) {
        // Sort + Three pointers
        // O(n^2) time & O(1) space

        if (nums.size() < 3) {
            return 0;
        }

        int n = nums.size();
        sort(nums.begin(), nums.end());

        int count = 0;
        for (int i = 0; i < n - 2; i++) {
            int j = i + 1, k = n - 1;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum < target) {
                    count += k - j;
                    j++;
                } else {
                    k--;
                }
            }
        }

        return count;
    }
};